contents/node/nodejs.md

```diff
@@ -62,31 +62,43 @@ Server running at http://127.0.0.1:3000/
+<figure>
+<img src="eventloop.png" width="74%" class="mx-auto mt-3 mb-2 d-block rounded" />
+<figcaption class="text-center"><small>
+
+Illustration of nodeJS's event loop from [these slides](https://www.slideshare.net/DevCentralAMD/introduction-to-nodejs-41469620)</small>
+</figcaption>
+</figure>
+
+Node is designed to be **event-driven**. When an <tooltip content="e.g. when a I/O operation is complete, or a timer fires">event</tooltip> occurs, the event handler previously registered with the event is queued to be run by the [event loop](https://medium.com/the-node-js-collection/what-you-should-know-to-really-understand-the-node-js-event-loop-and-its-metrics-c4907b19da4c). The main event loop (where Javascript code is executed) is single-threaded.
 
 This means that we can avoid thread overheads and synchronization problems including deadlocks and race conditions.
 
-### Benefit 3: Fast for I/O Intensive Tasks
+### Benefit 3: Fast for I/O Intensive Programs
+
+I/O requests made in other languages such as Python or Java are typically blocking, which means the program remains idle (in the same thread) until the I/O operation completes.
 
-This is due to the **non-blocking** I/O model.
+In contrast, nodeJS's allows for a **non-blocking** I/O model with its event-driven structure.
 
-When file and other I/O operations requests are made in Python or Java, they are blocking -- subsequent requests (in the same thread) have to be enqueued and can only be processed after the current operation and request completes. Meanwhile, the program remains idle. In contrast, Node has an event loop which delegates the actual tasks to other systems (e.g. file system and databases) and it is almost never blocked after delegation. While an I/O operation is incomplete, the event loop can still process subsequent requests.
+I/O requests are delegated to other systems (e.g. file system and databases). While an I/O operation is incomplete, the event loop can still process subsequent requests. When the I/O request is complete, the handler registered with the request is scheduled to be executed on the event loop again.
 
 ### Benefit 4: Use JavaScript for Both Front and Back-end Development
 
-Using Node for back-end development makes it possible to share and reuse code between the front-end and back-end. Teams can be more efficient and less reliant on documentation as they can read the code directly. The team can also become more cross-functional as developers can contribute to both development.
+Using Node for back-end development makes it possible to share common code functionality between the front-end and back-end, which leads to less code maintainence. It also decreases the potential communication overhead between front and back-end developers, since they share the same code base, and may hence also lead to a more multidisciplinary team.
```

Comment:
Consider replacing "It also decreases the potential..." with "It also reduces the potential..."

Consider appending the following after "... a more multidisciplinary team."
"that is familiar with both front and back-end development."